### DISCUSSION TITLE
Access git commit message

### DISCUSSION THREAD

**Discussion by xplosionmind** [11ty/eleventy#2324] [Q&A]
Hi everybody. I intend to place a “latest changes” section at the bottom of my website homepage. It should consist in a list of the latest 5 pages I updated, and I would like to display the commit message, like this:

```liquid
{% assign latestchanges = collections.all | sort: 'updated' | reverse %}
<ul>{% for page in latestchanges limit: 5 %}
  <li>
    {{ page.data.updated }} <b>{{ page.data.title }}</b> - {{ page.data.git_commit_msg }}
  </li>
{% endfor %}</ul>
```

How would I get the **`git_commit_msg`** variable?

Thanks a lot!

**Comment by pdehaan** [comment]
How/where are you setting the `git_commit_msg`? Is it in your front matter?

**Reply by pdehaan** [reply]
If you're asking in general how you can fetch the data from your Git history, you can use something like [**simple-git**](http://npm.im/simple-git) to call `git log`.

```js
// .eleventy.js
const simpleGit = require("simple-git");

module.exports = function (eleventyConfig) {
  const git = simpleGit();

  eleventyConfig.addFilter(
    "git_commit_msg",
    async function (file = this.page.inputPath) {
      const { latest } = await git.log({ file, maxCount: 1 });
      return latest.message;
    }
  );

  eleventyConfig.addShortcode(
    "git_commit_msg",
    async function (file = this.page.inputPath) {
      const { latest } = await git.log({ file, maxCount: 1 });
      // Make sure the file was found in git history (ie, not a newly created file which hasn't been committed yet.
      if (!latest) {
        return `<!-- git commit history not found for "${file}" -->`;
      }
      return `<span data-file="${file}" data-hash="${latest.hash}" data-date="${latest.date}">${latest.message}</span>`;
    }
  );

  return {
    dir: {
      input: "src",
      output: "www",
    },
  };
};
```

Then in your ./src/about.liquid file you could use the custom shortcode, which would display the latest git commit message for the current file:

```liquid
---
# ./src/about.liquid
title: About page
updated: "2022-02-22"
---

<h1>{{ title }}</h1>

{% git_commit_msg %}
```

Or from your example, you can use the custom `{% git_commit_msg %}` shortcode in the `for` loop and pass in the collection item's `inputPath` to get the latest commit info for a specific file:

```liquid
---
title: Index page
updated: "2022-01-31"
---

<h1>{{ title }}</h1>

{% assign latestchanges = collections.all | sort: 'updated' | reverse %}
<ul>
{%- for p in latestchanges limit: 5 %}
  <li>
    {{ p.data.updated }} <b>{{ p.data.title }}</b> - {% git_commit_msg p.inputPath %}
  </li>
{%- endfor %}
</ul>

{% git_commit_msg %}
```

Here's the output from my simple 3 page test site:

```html
<h1>Index page</h1>

<ul>
  <li>
    2022-03-13 <b>Contact page</b> - <span data-file="./src/contact.liquid" data-hash="6582a535cdb6465e117446c4c78c8b9029b3dfb2" data-date="2022-04-13T21:11:36-07:00">Update contact page to use shortcode instead of filter</span>
  </li>
  <li>
    2022-01-31 <b>Index page</b> - <span data-file="./src/index.liquid" data-hash="1e1973e200b310ab472199868dcaf0d7c2e4e03b" data-date="2022-04-13T21:12:06-07:00">Add git info to homepage</span>
  </li>
  <li>
    2022-02-22 <b>About page</b> - <span data-file="./src/about.liquid" data-hash="4840c88221340e588e8567673b80ca52fb72d56e" data-date="2022-04-13T21:11:15-07:00">Update about page to use shortcode</span>
  </li>
</ul>

<span data-file="./src/index.liquid" data-hash="1e1973e200b310ab472199868dcaf0d7c2e4e03b" data-date="2022-04-13T21:12:06-07:00">Add git info to homepage</span>
```

**Reply by xplosionmind** [reply]
> How/where are you setting the `git_commit_msg`? Is it in your front matter?

That is exactly the problem, @pdehaan… it should be retrieved by git *after* I committed the change, since it is the comment that describes the latest change I performed to the file.
Your subsequent examples make a lot of sense, thanks a lot!

Nevertheless, I still have two (unrelated) problems:
1. it appears that the `sort: 'updated'` filter does not work… dates appear quite randomly. what am I doing wrong?
2. the `git_commit_msg` for the latest global commit is continuously printed, not the one referred to the file!

Please refer to [this part of my source code](https://github.com/xplosionmind/tommi.space/blob/c28d519f56034cb6dfee123496521d97bcdd310f/content/home.html#L55).

Also, note that [here](https://github.com/xplosionmind/tommi.space/blob/c28d519f56034cb6dfee123496521d97bcdd310f/content/content.11tydata.js#L26) I made `updated` correspond to `date` in case there is no value.

**Reply by xplosionmind** [reply]
Ok, about point 2 it is my fault, I forgot to pass `inputPath`.

**Reply by pdehaan** [reply]
You might need `sort: 'data.updated'`:

```liquid
---
title: home
---

{% assign latestchanges = collections.pages | sort: 'data.updated' | reverse %}
<ul>
{% for page in latestchanges limit: 4 %}
  <li>
    {{ page.data.updated }} <b>{{ page.data.title }}</b>
  </li>
{% endfor %}
</ul>
```
